pynxtools/dataconverter/readers/stm/stm_file_parser.py

```diff
@@ -0,0 +1,383 @@
+
+            # In case if value contain name and unit e.g. /.../demodulated_signal: 'current(A)'
+            if start_bracket in val and end_bracket in val:
+                unit_parts = val.rsplit(start_bracket)
+                # Assume that val does not have any key but decriptive text,
+                # e.g. Current (A);Bias (V);
+                if len(unit_parts) > 2:
+                    return [(key, val)]
+                tmp_l_part, tmp_r_part = unit_parts
+                unit = tmp_r_part.rsplit(end_bracket)[0]
+                val = tmp_l_part.strip()
+                if unit in UNIT_TO_SKIP:
+                    unit = ''
+                return [(key, val), (f"{key}/@unit", unit)]
+
+        return []
+
+    def get_sxm_raw_metadata_and_signal(self, file_name):
+        """
+        Retun metadata plain dict and signal
+        Convert header part (that contains metadata) of a file with 'sxm' extension into
+        plain dict.
+        """
+        scan_file = nap.read.Scan(file_name)
+        header_end_byte = scan_file.start_byte()
+        h_part = scan_file.read_raw_header(header_end_byte)
+        while True:
+            # Ignore all starting chars of string h_part except Alphabat
+            if not re.match("[a-zA-Z]", h_part):
+                h_part = h_part[1:]
+            else:
+                break
+
+        h_comp_iter = iter(re.split('\n:|:\n', h_part))
+        return dict(zip(h_comp_iter, h_comp_iter)), scan_file.signals
+
+    def get_SPM_metadata_dict_and_signal(self):
+        """
+        Get meradata and signal from spm file.
+        """
+        metadata_dict, signal = self.get_sxm_raw_metadata_and_signal(self.file_name)
+        nesteded_matadata_dict = self.get_nested_dict_from_concatenated_key(metadata_dict)
+        # Convert nested (dict) path to signal into slash_separated path to signal
+        temp_flattened_dict_sig = {}
+        nested_path_to_slash_separated_path(signal,
+                                            temp_flattened_dict_sig)
+        temp_flattened_dict = {}
+        nested_path_to_slash_separated_path(nesteded_matadata_dict,
+                                            temp_flattened_dict)
+        flattened_dict = {}
+        for key, val in temp_flattened_dict.items():
+            # list of tuples of (data path, data) and (unit path/unit and unit value)
+            tuple_li = self.convert_key_to_unit_and_entity(key, val,
+                                                           start_bracket='(',
+                                                           end_bracket=')')
+            if tuple_li:
+                for tup in tuple_li:
+                    flattened_dict[tup[0]] = tup[1]
+            else:
+                flattened_dict[key] = val
+
+        flattened_dict.update(temp_flattened_dict_sig)
+        return flattened_dict
+
+    # pylint: disable=too-many-arguments
+    def construct_nxdata_for_sxm(self,
```

Comment:
I have a general question about this design here as I can pick it up too quickly from the code here:
In general for each file format e.g. nanonis or sxm you have on the left hand side some information
which you want to map on a NeXus concept. What I don't see immediately transparent from the code
where is the logic which decides which of the fields on the left hand side will for sure be copied over and which optionally.
In other words is the code logic reasonably stable to deal with changes of the NXsts appdef when in this appdef required and optional constraints are changed. It seems to me that one strength of your implementation is the fact that you do not directly hardcode paths but try to resolve many paths programmatically which is good.

Reply:
This parse only read the numerical data and arrays, collects the units, long-name, that's it. And it fills the concept with the corresponding data. To check about the concepts of optional and required fields passed the converter validation steps. The file parser or stm reader does not have anything to do with required and optional concepts.